### PR TITLE
#86 Fix singletone class initialisation

### PR DESCRIPTION
**Pull request by Romick2005**
Also take care of start stop Observing issue on metro bundler restart/reload/hot-reload.

### CODE REVIEW

**Comment by zxcpoiu** [comment]
@Romick2005 

So you do not have init multiple times after this PR?

**Comment by Romick2005** [comment]
Actually init called multiple times, but initPrivate only ones. Also on multiple call of init the returning result is same initialised only one time object. But generally this PR implements real singletone pattern.

**Comment by zxcpoiu** [comment]
About `startObserving` and `stopObserving`, would you please remove them and treat it as your own `workaround` in your private branch?

I don't think `to ignore stopObserving when debug`  is a valid solution here.

`Reload` process has causes similar issues around different libs ( like mis-created multiple instances for react-native-webrtc, IIRC ), given that the reload only happens on DEBUG mode only, and we can not assure that the behavior is a bug or whether it will change in the future or not, not to say there are different algorithms / fallbacks / Engines there ( ex: full reload / HMS Fast Refresh...etc ), I prefer not to add this for general usages. Personally, I only find Fast Refresh useful when I'm tweaking UI stuff, if the test is related to native library, I always did a clean restart ( swipe out ) to prevent any reload-caused issues.

I think the more appropriate way would be needing an event api like `onRefresh` or `onReload` for the native side provided by React Native, and user can config them to reset some states.

**Review by zxcpoiu — CHANGES_REQUESTED** [review]
1. remove `stopObserving` workaround for refresh, keep this PR simple and could be discussed in another PR

2. remove unused / commented debugging logs

File: ios/RNVoipPushNotification/RNVoipPushNotificationManager.m [hunk 1]
```diff
@@ -90,12 +95,18 @@ - (void)startObserving
 {
     _hasListeners = YES;
     if ([_delayedEvents count] > 0) {
+        // RCTLog(@"[RNVoipPushNotificationManager] sendEventWithName _hasListeners = %d name = %@", _hasListeners, RNVoipPushDidLoadWithEvents);
```
Comment (zxcpoiu) [inline ios/RNVoipPushNotification/RNVoipPushNotificationManager.m:98]: remove debugging comments

File: ios/RNVoipPushNotification/RNVoipPushNotificationManager.m [hunk 2]
```diff
@@ -90,12 +95,18 @@ - (void)startObserving
...
     }
 }
 
 - (void)stopObserving
 {
+#ifdef DEBUG
+    // RCTLog(@"[RNVoipPushNotificationManager] skip stopObserving because in debug mode startObserving is not called on reload, but stopObserving is called before reload");
+    return;
+#endif
+
```
Comment (zxcpoiu) [inline ios/RNVoipPushNotification/RNVoipPushNotificationManager.m:109]: remove them or start another PR to continue discussion

File: ios/RNVoipPushNotification/RNVoipPushNotificationManager.m [hunk 3]
```diff
@@ -107,6 +118,7 @@ - (void)stopObserving
 
 // --- send directly if has listeners, cache it otherwise
 - (void)sendEventWithNameWrapper:(NSString *)name body:(id)body {
+    // RCTLog(@"[RNVoipPushNotificationManager] sendEventWithNameWrapper _hasListeners = %d  name = %@", _hasListeners, name);
```
Comment (zxcpoiu) [inline ios/RNVoipPushNotification/RNVoipPushNotificationManager.m:121]: remove debugging comments

File: ios/RNVoipPushNotification/RNVoipPushNotificationManager.m [hunk 4]
```diff
@@ -30,27 +30,32 @@ @implementation RNVoipPushNotificationManager
...
+    return [[self class] sharedInstance];
+}
 
-- (instancetype)init
-{
-    if (self = [super init]) {
+-(instancetype) initPrivate {
+    self = [super init];
+    if (self) {
+        // Initialization code here.
```
Comment (zxcpoiu) [inline ios/RNVoipPushNotification/RNVoipPushNotificationManager.m:44]: above should only change the function name, instead making a mess diff

**Comment by zxcpoiu** [comment]
@jonastelzio @danjenkins 

Hello, would you mind to take a look if this solves your problem described in #85?

File: ios/RNVoipPushNotification/RNVoipPushNotificationManager.m [hunk 5]
```diff
@@ -30,27 +30,28 @@ @implementation RNVoipPushNotificationManager
...
 
-- (instancetype)init
-{
-    if (self = [super init]) {
-        _delayedEvents = [NSMutableArray array];
-    }
-    return self;
+-(instancetype) initPrivate {
+    _delayedEvents = [NSMutableArray array];
+    return [super init];
```
Comment (zxcpoiu) [inline ios/RNVoipPushNotification/RNVoipPushNotificationManager.m:43]: wouldn't it be:

```objc
- (instancetype)initPrivate
{
    if (self = [super init]) {
        _delayedEvents = [NSMutableArray array];
    }
    return self;
}
```

?

**Review by zxcpoiu — CHANGES_REQUESTED** [review]

File: ios/RNVoipPushNotification/RNVoipPushNotificationManager.m (hunk 5 above)
Comment (zxcpoiu) [reply ios/RNVoipPushNotification/RNVoipPushNotificationManager.m:43]: The diff might looks like this:

```diff
- (instancetype)init
+ (instancetype)initPrivate
```

all looks good now except this.

Comment (Romick2005) [reply ios/RNVoipPushNotification/RNVoipPushNotificationManager.m:43]: There are two methods init and initPrivate. Init should be there as it is inherited from base class and it invocation will return allocated and initialised only once with initPrivate method.
It is based on https://stackoverflow.com/a/7035136/3686678

Comment (Romick2005) [reply ios/RNVoipPushNotification/RNVoipPushNotificationManager.m:43]: > wouldn't it be:
> 
> ```objc
> - (instancetype)initPrivate
> {
>     if (self = [super init]) {
>         _delayedEvents = [NSMutableArray array];
>     }
>     return self;
> }
> ```
> 
> ?

Actually initPrivate is called only once so it should be ok.

Comment (zxcpoiu) [reply ios/RNVoipPushNotification/RNVoipPushNotificationManager.m:43]: yeh, I know, I just want to keep it as less diff as possible, and not cause too much confuses for the future developers

Comment (Romick2005) [reply ios/RNVoipPushNotification/RNVoipPushNotificationManager.m]: Yep, I see. But I also prefer to improve code ). And for me is not appropriate to make an assignment inside condition. It is unusual and hard to read. I would better go with extra line for  assignment and the check in if condition.

**Comment by zxcpoiu** [comment]
Thanks @Romick2005 
I just fix the style to match with the current code base, sorry to push to your branch directly, but I think it's more efficient.
If you want to adjust / fix the style, feel free to send another PR, thanks!

**Comment by zxcpoiu** [comment]
Since this an open source, so it would be better to follow the current code base's style, and makes minimal changes in a single PR.

But fire a PR to adjust them is acceptable for sure

**Comment by Romick2005** [comment]
Thank you @zxcpoiu, it's just styling, so never mind. No worries I am ok with this, until it works as expected :).

**Pull request merged**